cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java

```diff
@@ -187,13 +189,56 @@ public void removeMetadata(Id.NamespacedId targetId, Predicate<String> filter) {
    * Find the instance of {@link BusinessMetadataRecord} based on key.
    *
    * @param value The metadata value to be found.
-   * @return The {@Iterable} of {@link BusinessMetadataRecord} that fit the key.
+   * @param type The target type of objects to search from.
+   * @return The {@link Iterable} of {@link BusinessMetadataRecord} that fit the value.
    */
-  public Iterable<BusinessMetadataRecord> findBusinessMetadataOnValue(String value) {
+  public List<BusinessMetadataRecord> findBusinessMetadataOnValue(String value, MetadataSearchTargetType type) {
+    return executeSearchOnColumns(BusinessMetadataDataset.VALUE_COLUMN, value, type);
+  }
+
+  /**
+   * Find the instance of {@link BusinessMetadataRecord} for key:value pair
+   *
+   * @param keyValue The metadata value to be found.
+   * @param type The target type of objects to search from.
+   * @return The {@link Iterable} of {@link BusinessMetadataRecord} that fit the key value pair.
+   */
+  public List<BusinessMetadataRecord> findBusinessMetadataOnKeyValue(String keyValue, MetadataSearchTargetType type) {
+    return executeSearchOnColumns(BusinessMetadataDataset.KEYVALUE_COLUMN, keyValue, type);
+  }
+
+  // Helper method to execute IndexedTable search on target index column.
+  List<BusinessMetadataRecord> executeSearchOnColumns(String column, String searchValue,
```

Comment:
I set it as package private for now just in case we want to access it to do unit tests.